### PR TITLE
Cleanup Brave Ads transfer/tab manager observers

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/24644

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Confirm ads are landed (aka transferred) after 10 seconds

### CODE REVIEW

**Review by kylehickinson — CHANGES_REQUESTED** [review]
You didn't update `brave_ads.h` where the method is declared to match changed signature in `brave_ads.mm`

File: ios/browser/api/ads/brave_ads.mm [hunk 1]
```diff
@@ -567,14 +567,19 @@ - (void)reportMediaStoppedWithTabId:(NSInteger)tabId {
...
   if (![self isAdsServiceRunning]) {
     return;
   }
-  ads->OnTabUpdated((int32_t)tabId, net::GURLWithNSURL(url), isSelected,
-                    [self isBrowserActive], isPrivate);
+  std::vector<GURL> urls;
+  for (NSURL* redirectURL in redirectionURLs) {
+    urls.push_back(net::GURLWithNSURL(redirectURL));
+  }
+  urls.push_back(net::GURLWithNSURL(url));
```
Comment (aseren) [inline ios/browser/api/ads/brave_ads.mm:580]: I don't see the url variable here

File: vendor/bat-native-ads/src/bat/ads/internal/processors/behavioral/purchase_intent/purchase_intent_processor.cc [hunk 2]
```diff
@@ -224,7 +224,7 @@ void PurchaseIntent::OnResourceDidUpdate(const std::string& id) {
 }
 
 void PurchaseIntent::OnTextContentDidChange(
-    const int32_t id,
+    const int32_t tab_id,
     const std::vector<GURL>& redirect_chain,
     const std::string& content) {
   const GURL& url = redirect_chain.back();
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/processors/behavioral/purchase_intent/purchase_intent_processor.cc:230]: we need to check redirect_chain size before calling back() method

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/processors/behavioral/purchase_intent/purchase_intent_processor.cc:230]: The redirect chain is always guaranteed to have at least 1 entry. See `navigation_handle->GetRedirectChain()`

File: ios/browser/api/ads/brave_ads.mm (hunk 1 above)
Comment (kylehickinson) [inline ios/browser/api/ads/brave_ads.mm:580]: You deleted the `url` argument, so this line has to be removed or it won't compile. If you want this API to mimic `reportLoadedPageWithURL` make sure to keep the `url` argument

**Review by kylehickinson — CHANGES_REQUESTED** [review]

File: ios/browser/api/ads/brave_ads.mm (hunk 1 above)
Comment (tmancey) [reply ios/browser/api/ads/brave_ads.mm:581]: Fixed, thanks

**Review by kylehickinson — APPROVED** [review]
iOS++

File: vendor/bat-native-ads/src/bat/ads/internal/processors/behavioral/purchase_intent/purchase_intent_processor.cc (hunk 2 above)
Comment (aseren) [reply vendor/bat-native-ads/src/bat/ads/internal/processors/behavioral/purchase_intent/purchase_intent_processor.cc:230]: in this case we need to add DCHECK here

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/processors/behavioral/purchase_intent/purchase_intent_processor.cc:232]: Added, thanks

File: vendor/bat-native-ads/src/bat/ads/internal/processors/behavioral/purchase_intent/purchase_intent_processor.cc [hunk 3]
```diff
@@ -237,8 +239,12 @@ void PurchaseIntent::OnTextContentDidChange(
 
   const absl::optional<TabInfo> last_visible_tab =
       TabManager::GetInstance()->GetLastVisibleTab();
-  if (SameDomainOrHost(url,
-                       last_visible_tab ? last_visible_tab->url : GURL())) {
+  if (!last_visible_tab) {
+    return;
+  }
+
+  DCHECK(!redirect_chain.empty());
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/processors/behavioral/purchase_intent/purchase_intent_processor.cc:246]: should be `DCHECK(last_visible_tab->redirect_chain.back());`

Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/processors/behavioral/purchase_intent/purchase_intent_processor.cc]: Fixed

File: vendor/bat-native-ads/src/bat/ads/internal/transfer/transfer.cc [hunk 4]
```diff
@@ -91,14 +91,14 @@ void Transfer::OnTransferAd(const int32_t tab_id,
...
-      TabManager::GetInstance()->GetTabForId(tab_id);
-  if (!tab) {
+  absl::optional<TabInfo> tab = TabManager::GetInstance()->GetTabForId(tab_id);
+  if (!tab || tab->redirect_chain.empty()) {
     FailedToTransferAd(ad);
     return;
   }
 
-  if (!DomainOrHostExists(redirect_chain, tab->url)) {
+  DCHECK(!redirect_chain.empty());
```
Comment (aseren) [inline vendor/bat-native-ads/src/bat/ads/internal/transfer/transfer.cc:100]: this should be DCHECK(!tab->redirect_chain.empty());

**Review by aseren — APPROVED** [review]
Approved with one comment

File: vendor/bat-native-ads/src/bat/ads/internal/transfer/transfer.cc (hunk 4 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/transfer/transfer.cc:102]: Fixed

**Pull request merged**